compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -0,0 +1,329 @@
+            .functions
+            .iter_mut()
+            .filter(|(_, func)| matches!(func.runtime(), RuntimeType::Brillig(_)));
+
+        for (_, function) in brillig_functions {
+            function.loop_invariant_code_motion();
+        }
+
+        self
+    }
+}
+
+impl Function {
+    fn loop_invariant_code_motion(&mut self) {
+        Loops::find_all(self).hoist_loop_invariants(self);
+    }
+}
+
+impl Loops {
+    fn hoist_loop_invariants(self, function: &mut Function) {
+        let mut inserter = FunctionInserter::new(function);
+
+        for loop_ in self.yet_to_unroll {
+            let Ok(pre_header) = loop_.get_pre_header(inserter.function, &self.cfg) else {
+                // If the loop does not have a preheader we skip hoisting loop invariants for this loop
+                continue;
+            };
+
+            // Gather the variables declared within the loop
+            let mut defined_in_loop: HashSet<ValueId> = HashSet::default();
+            for block in loop_.blocks.iter() {
+                let params = inserter.function.dfg.block_parameters(*block);
+                defined_in_loop.extend(params);
+                for instruction_id in inserter.function.dfg[*block].instructions() {
+                    let results = inserter.function.dfg.instruction_results(*instruction_id);
+                    defined_in_loop.extend(results);
+                }
+            }
+
+            // Instructions to move to the preheader
+            let mut instructions_to_hoist = Vec::new();
+            // Mapping to track unchanged instructions per block
+            let mut block_to_instructions = HashMap::default();
+            // Track already found loop invariants
+            let mut loop_invariants = HashSet::default();
+            for block in loop_.blocks.iter() {
+                let mut instructions_to_keep = Vec::new();
+                for instruction_id in inserter.function.dfg[*block].take_instructions() {
+                    let mut is_not_loop_invariant = false;
+                    // The list of blocks for a nested loop contain any inner loops as well.
+                    // We may have already re-inserted new instructions if two loops share blocks
+                    // so we need to map all the values in the instruction which we want to check.
+                    let (instruction, _) = inserter.map_instruction(instruction_id);
+                    instruction.for_each_value(|value| {
+                        // We are implicitly checking whether the values are constant as well.
+                        // The set of values defined in the loop only contains instruction results and block parameters
+                        // which cannot be constants.
+                        is_not_loop_invariant |=
+                            defined_in_loop.contains(&value) && !loop_invariants.contains(&value);
+                    });
+
+                    if !is_not_loop_invariant
```

Comment:
```suggestion
                    if is_loop_invariant
```
Double-negated boolean (`!` and `_not_`)

Reply:
Switched to not do the final double negation. I had to change the logic of the instruction value mapping above as well.